src/wallet/gtest/test_rpc_wallet.cpp

```diff
@@ -239,3 +240,163 @@ TEST(WalletRPCTests, RPCZsendmanyTaddrToSapling)
+        CScript scriptPubKey = CScript() << OP_DUP << OP_HASH160 << ToByteVector(taddr) << OP_EQUALVERIFY << OP_CHECKSIG;
+        size_t utxoCount = 100;
+        for (size_t i = 0; i < utxoCount; i++) {
+            mtx.vout.push_back(CTxOut(5 * COIN, scriptPubKey));
+        }
+        CWalletTx wtx(pwalletMain, mtx);
+        pwalletMain->LoadWalletTx(wtx);
+
+        // Fake-mine the transaction
+        EXPECT_EQ(-1, chainActive.Height());
+        CBlock block;
+        block.vtx.push_back(wtx);
+        block.hashMerkleRoot = BlockMerkleRoot(block);
+        auto blockHash = block.GetHash();
+        CBlockIndex fakeIndex {block};
+        mapBlockIndex.insert(std::make_pair(blockHash, &fakeIndex));
+        chainActive.SetTip(&fakeIndex);
+        EXPECT_TRUE(chainActive.Contains(&fakeIndex));
+        EXPECT_EQ(0, chainActive.Height());
+        wtx.SetMerkleBranch(block);
+        pwalletMain->LoadWalletTx(wtx);
+
+        // Add keys manually
+        std::string taddr_string = keyIO.EncodeDestination(taddr);
+
+        WalletTxBuilder builder(Params(), minRelayTxFee);
+
+        auto selector = CWallet::LegacyTransparentZTXOSelector(
+                true,
+                TransparentCoinbasePolicy::Disallow);
+
+        { // test transparent inputs to NetAmountRecipient
+            auto saplingKey = pwalletMain->GenerateNewLegacySaplingZKey();
+            NetAmountRecipient zaddr2(saplingKey, std::nullopt);
+
+            TransactionStrategy strategy(PrivacyPolicy::AllowRevealedSenders);
+
+            SpendableInputs inputs;
+            for (size_t i = 0; i < utxoCount; i++) {
+                inputs.utxos.emplace_back(COutput(&wtx, i, 100, true));
+            }
+
+            auto effects = builder.PrepareTransaction(
+                    *pwalletMain,
+                    selector,
+                    inputs,
+                    zaddr2,
+                    chainActive,
+                    strategy,
+                    std::nullopt,
+                    1)
+            .map_error([&](const auto& err) {
+                try {
+                    ThrowInputSelectionError(err, selector, strategy);
+                } catch (const UniValue& value) {
+                    FAIL() << value.write();
+                }
+            })
+            .value();
+
+            auto buildResult = effects.ApproveAndBuild(
+                    Params().GetConsensus(),
+                    *pwalletMain,
+                    chainActive,
+                    strategy);
+            auto tx = buildResult.GetTxOrThrow();
+
+            auto expectedFee = tx.GetConventionalFee();
+            // Leave one incremental fee tick of buffer above the conventional fee.
+            EXPECT_TRUE(effects.GetFee() == expectedFee
+                        || effects.GetFee() == expectedFee + 5000)
+            << "effects.GetFee() = " << effects.GetFee() << std::endl
+            << "expectedFee = " << expectedFee;
```

Comment:
This could be a closure taking the output of `builder.PrepareTransaction` that is shared between the two parts of the test. Non-blocking.